### ISSUE TITLE
Missing memcached_exists symbol

### ISSUE THREAD

**Issue by mvanbaak** [awslabs/aws-elasticache-cluster-client-memcached-for-php#12]
2 days ago upstream memcached module sources were pulled into the php7 branch. (commit b55106b26b78907a1b2b6b9813e645c7fe978ffc)

With this commit the module cannot be loaded anymore:

```
PHP Warning:  PHP Startup: Unable to load dynamic library '/usr/lib64/php/7.0/modules/memcached.so' - /usr/lib64/php/7.0/modules/memcached.so: undefined symbol: memcached_exist in Unknown on line 0
```

**Comment by mvanbaak** [comment]
Probably fixed as soon as you guys sync https://github.com/awslabs/aws-elasticache-cluster-client-libmemcached with latest upstream.

**Comment by sodabrew** [comment]
Hmm. I thought `memcached_exist` was available in all libmemcached 1.0.x versions.

**Comment by mvanbaak** [comment]
I thought so as well, but if you look at the commit where the upstream changes are pulled, there's an ifdef for that function in php-memcached-compat. This is probably the change that breaks it (the removal of this whole c and h file).

All documents state that libmemcached 1.X is ok, but it seems the latest 3.0 version needs a newer libmemcached. 

I'm not familiar with libmemcached and the php module, so this is purely a 'I think' solution without looking into the code of libmemcached itself ;-)

**Comment by sodabrew** [comment]
libmemcached Changelog:
```
0.53 Mon Sep 26 20:50:33 PDT 2011
  * Fix for FreeBSD/OpenBSD and -lm
  * Added memcached_exist()
  * Fix for memory when using config test.
  * CLI gained --quiet
```

**Comment by mvanbaak** [comment]
I launched a clean amazon linux ami (the latest) to compile the module.
This is my ansible setup for it:

```
- name: Clone aws-elasticache-cluster-client-libmemcached library
  tags: awsmemcached
  git:
    repo: https://github.com/awslabs/aws-elasticache-cluster-client-libmemcached.git
    dest: /tmp/aws-libmemcached
    accept_hostkey: yes
    depth: 1

- name: Create build directory
  tags: awsmemcached
  file:
    dest: /tmp/aws-libmemcached/BUILD
    mode: 0755
    state: directory

- name: Configure aws-elasticache-cluster-client-libmemcached library
  tags: awsmemcached
  shell: >
    ../configure --prefix=/usr/local/lib/aws-libmemcached --with-pic
    chdir=/tmp/aws-libmemcached/BUILD

- name: Make and install aws-elasticache-cluster-client-libmemcached library
  tags: awsmemcached
  shell: >
    {{ item }}
    chdir=/tmp/aws-libmemcached/BUILD
  with_items:
    - make
    - make install

- name: Clone aws-elasticache-cluster-client-memcached-for-php
  tags: awsmemcached
  git:
    repo: https://github.com/awslabs/aws-elasticache-cluster-client-memcached-for-php.git
    dest: /tmp/aws-elasticache-cluster-client-memcached-for-php
    accept_hostkey: yes
    depth: 1
    version: php7

- name: Configure, compile and install aws-elasticache-cluster-client-memcached-for-php
  tags: awsmemcached
  shell: >
    {{ item }}
    chdir=/tmp/aws-elasticache-cluster-client-memcached-for-php
  with_items:
    - phpize
    - ./configure --with-libmemcached-dir=/usr/local/lib/aws-libmemcached --disable-memcached-sasl --enable-memcached-igbinary
    - sed -i "s#-lmemcached#/usr/local/lib/aws-libmemcached/lib/libmemcached.a -lcrypt -lpthread -lm -lstdc++#" Makefile
    - make
    - make install

- name: Enable aws-elasticache-cluster-client-memcached-for-php
  tags: php
  lineinfile:
    dest: "/etc/{{ php_conf_dir }}/zz-memcached.ini"
    line: "extension=memcached.so"
    create: yes
```

With the latest commit this gives the error `undefined symbol: memcached_exist`

When I checkout `7937c7f1907f74dc857ed8f34a0261bf0de89ae2` instead of `HEAD` of the php7 branch everything works

**Comment by malyeyev-AMZN** [comment]
Relevant changes in upsteam libmemcached between 1.0.8 and 1.0.18:

diff --git a/libmemcached-1.0/exist.h b/libmemcached-1.0/exist.h
index 5c60105..ef60a81 100644
--- a/libmemcached-1.0/exist.h
+++ b/libmemcached-1.0/exist.h
@@ -36,6 +36,10 @@
 
 #pragma once
 
+#ifdef __cplusplus
+extern "C" {
+#endif
+
 LIBMEMCACHED_API
 memcached_return_t memcached_exist(memcached_st *memc, const char *key, size_t key_length);
 
@@ -43,3 +47,6 @@ LIBMEMCACHED_API
 memcached_return_t memcached_exist_by_key(memcached_st *memc,
                                           const char *group_key, size_t group_key_length,
                                           const char *key, size_t key_length);
+#ifdef __cplusplus
+}
+#endif

**Comment by sodabrew** [comment]
Is this a bug in libmemcached for visibility of that symbol to C linkage?

**Comment by malyeyev-AMZN** [comment]
@mvanbaak do you by any chance have multiple versions of libmemcached installed?
`sudo yum list installed | grep libmem`

**Comment by malyeyev-AMZN** [comment]
Still can't reproduce it.

Unlikely but this may also be related to libmemcachedutil library. Just to rule it out, could you please try (between running ./configure and make) using the following

`sed -i "s#-lmemcached\b#/usr/local/lib/aws-libmemcached/lib/libmemcached.a -lcrypt -lpthread -lm -lstdc++#" Makefile && sed -i "s#-lmemcachedutil\b#/usr/local/lib/aws-libmemcached/lib/libmemcachedutil.a#" Makefile`
instead of
`sed -i "s#-lmemcached#/usr/local/lib/aws-libmemcached/lib/libmemcached.a -lcrypt -lpthread -lm -lstdc++#" Makefile`

**Comment by mvanbaak** [comment]
I'll test that later tonight. Sorry, bit busy at the moment

**Comment by malyeyev-AMZN** [comment]
@mvanbaak 
Also I can't run your playbook
```
ansible-playbook -i "localhost," -c local issue-12.yml
ERROR! 'git' is not a valid attribute for a Play

The error appears to have been in '/home/ec2-user/issue-12.yml': line 1, column 3, but may
be elsewhere in the file depending on the exact syntax problem.

The offending line appears to be:


- name: Clone aws-elasticache-cluster-client-libmemcached library
  ^ here
```

**Comment by malyeyev-AMZN** [comment]
Ok, able to reproduce now and debugging it.

**Comment by malyeyev-AMZN** [comment]
@sodabrew looks like one of the changes upstream broke it:

https://github.com/php-memcached-dev/php-memcached/commit/c564fd875ead3110b27dd3e1dfb9e5d4afabd042#diff-9441255ac664d11fa292d9fb7ce03320L24

```
]$ nm /tmp/aws-elasticache-cluster-client-memcached-for-php/modules/memcached.so | grep memcached_exist
                 U memcached_exist
]$ nm /usr/local/lib/aws-libmemcached/lib/libmemcached.a | grep memcached_exist
00000000000003f5 T _Z15memcached_existP12memcached_stPKcm
000000000000042a T _Z22memcached_exist_by_keyP12memcached_stPKcmS2_m
```

**Comment by mvanbaak** [comment]
Ok. If you want me to test something let me know. For now I told my playbook to use an older commit from the php7 branch ;-)

**Comment by malyeyev-AMZN** [comment]
For reference I'm adding full Ansible YAML source and instructions to reproduce this which also happen to be modified to not use dynamic linking (produces same outcome as in original setup).

To run:
```
ansible-playbook -i "localhost," -c local issue-12-shared-lib.yml
```

To clean up:
```
sudo rm -rf /tmp/aws-* /etc/php-7.0.d/zz-memcached.ini /etc/ld.so.conf.d/aws-libmemcached.conf
```

issue-12-shared-lib.yml:
```
- hosts: localhost
  sudo: True
  tasks:
  - name: Clone aws-elasticache-cluster-client-libmemcached library
    tags: awsmemcached
    git:
      repo: https://github.com/awslabs/aws-elasticache-cluster-client-libmemcached.git
      dest: /tmp/aws-libmemcached
      accept_hostkey: yes
      depth: 1
  
  - name: Create build directory
    tags: awsmemcached
    file:
      dest: /tmp/aws-libmemcached/BUILD
      mode: 0755
      state: directory
  
  - name: Configure aws-elasticache-cluster-client-libmemcached library
    tags: awsmemcached
    shell: >
      ../configure --prefix=/usr/local/lib/aws-libmemcached --with-pic
      chdir=/tmp/aws-libmemcached/BUILD
  
  - name: Make and install aws-elasticache-cluster-client-libmemcached library
    tags: awsmemcached
    shell: >
      {{ item }}
      chdir=/tmp/aws-libmemcached/BUILD
    with_items:
      - make
      - make install
  
  - name: Add libmemcached isntall directory to system search path for shared libraries
    shell:
      echo "/usr/local/lib/aws-libmemcached/lib" > /etc/ld.so.conf.d/aws-libmemcached.conf
  
  - name: Clone aws-elasticache-cluster-client-memcached-for-php
    tags: awsmemcached
    git:
      repo: https://github.com/awslabs/aws-elasticache-cluster-client-memcached-for-php.git
      dest: /tmp/aws-elasticache-cluster-client-memcached-for-php
      accept_hostkey: yes
      depth: 1
      version: php7
  
  - name: Configure, compile and install aws-elasticache-cluster-client-memcached-for-php
    tags: awsmemcached
    shell: >
      {{ item }}
      chdir=/tmp/aws-elasticache-cluster-client-memcached-for-php
    with_items:
      - phpize
      - ./configure --with-libmemcached-dir=/usr/local/lib/aws-libmemcached --disable-memcached-sasl
      - make
      - make install
  
  - name: Enable aws-elasticache-cluster-client-memcached-for-php
    tags: php
    lineinfile:
      dest: "/etc/php-7.0.d/zz-memcached.ini"
      line: "extension=memcached.so"
      create: yes
```

**Comment by malyeyev-AMZN** [comment]
Discussion moved to https://github.com/php-memcached-dev/php-memcached/pull/295#issuecomment-276951587

Once it is resolved upstream we will be merge the fix back into this repo

**Comment by sodabrew** [comment]
I've tagged php-memcached 3.0.1 with the fix for memcached_exists and others:
https://github.com/php-memcached-dev/php-memcached/releases/tag/v3.0.1

**Comment by malyeyev-AMZN** [comment]
Thanks Aaron. I've merged 3.0.1 into our php7 branch.

**Comment by mvanbaak** [comment]
I just re-ran my ansible playbook on a completely new amazonlinux ami:

https://puu.sh/tVnhz/9050d46f28.png

Seems to be fixed. Thanks!
